### PR TITLE
Recode ID Values from CRM

### PR DESCRIPTION
**Pull request by SashaWeinstein**
# Recode ID Values
This PR is to address issue #25 which was opened when Matt Croswell noticed that the open datasets sent to digital ocean had some fields still populated with non-human readable values. The issue is that there are two types of integer representation in the CRM. 
The first type is values like `717170000` that are placeholders for values that are re-used many times throughout the CRM. Matt has a list of the mappings this [this spreadsheet](https://nyco365.sharepoint.com/:x:/r/sites/NYCPLANNING/itd/edm/_layouts/15/Doc.aspx?sourcedoc=%7B31588592-8910-4FAD-B77C-0C62E106C794%7D&file=Attributes.Values.for.EDM_FullList.xlsx&action=default&mobileredirect=true), and the code to apply those mappings was merged in PR #19. The mappings can be downloaded from API from one single query. 
The second type of values to recode are what I've termed "ID values" which look like `0014e50f-1133-e811-8126-1458d04dc8c8`. These are different in that they aren't re-used nearly as much, and the mappings for a given record come from an API query specific to that record. These values are re-used, so `0014e50f-1133-e811-8126-1458d04dc8c8` will map to the same string in another record. However many records have at least one unique ID value and thus need a specific API query run to get back the mappings.

## New recode process

Lives in a new module called recode_id. There are four fields populated with these ID values, which are hardcoded into a constant in the top of the module. 

### Iterate through records and fields 
Most records need their own API call, so I iterate through each record with `.apply`. The rest of the logic is in `recode_single_project`, where the code first uses a `ReuseTracker` instance to check if the ID's present in the record have already been mapped. Sometimes each ID value has been seen already and can be mapped from the `ReuseTracker` attributes. But usually a new url for the API is generated and then the mappings converted to a dictionary from the json. I think taking a random record and seeing the mappings returned by the API is helpful to understand the code. 

### Re-use mapping with `ReuseTracker`
Due to long runtimes this class was implemented to track the mappings seen in earlier records. Each new ID value/human readable value pair is added to the corresponding dictionary and then used to recode if the ID value is seen again in a later record. 

### Get fresh headers
From my observation, it seems that the headers used to authorize the request can only be used a limited number of times before the API returns a 401 permission denied. In this situation there is code to get new headers and try again, and if this doesn't work then give up and row as is. 

## Logging process
Need logs as run-time are long and I want to get info about a run while it's going

## New Workflow
Wanted a workflow to test my new code. The new ` Export Single Dataset` action runs and exports a single open dataset to DO. It was modeled off of a similar "single dataset" action in data library

## Refactor to simplify
As I added new code I needed to refactor to keep things organized. I refactored code to new functions in `visible_projects.py` and moved my new recode ID code to a new module.

## Actions with new output
Export [dcp_projects](https://github.com/NYCPlanning/db-zap-opendata/runs/6779552446?check_suite_focus=true) and [dcp_projectbbls](https://github.com/NYCPlanning/db-zap-opendata/actions/runs/2456379897). Outputs should be [here](https://cloud.digitalocean.com/spaces/edm-publishing?i=266877&path=db-zap%2F20220607%2F) on digital ocean  

## Clean `project_id` field in project bbls dataset
This was the second thing Matt mentioned in the issue, fortunately much easier, one line change of 
```python
df["project_id"] = df["project_id"].str.split(" ").str[0]
```
in `src/runner.py`

### CODE REVIEW

**Review by mbh329 — COMMENTED** [review]
Okay sorry this took me so long, I should have used the flag to just query a subset of the data. Everything was going smoothly and the code makes sense to me (the logs are clear and the values are being recoded with human readable names). I was unable to fully execute the `python3 -m src.runner dcp_projects`because of a `requests.exceptions.ConnectionError: ('Connection aborted.', OSError(50, 'Network is down'))` error. In your PR write up, you mention that the api was returning an error but this seems to be a different error. I take it that the crm api is a bit finicky but was curious if you had encountered this error.

File: src/recode_id.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+            "dcp_applicant_customer_account": ("name", "accountid"),
+            "dcp_applicant_customer_contact": ("fullname", "contactid"),
+        },
+    },
+    "ceqr_leadagency": {
+        "metadata_field_names": ["dcp_leadagencyforenvreview"],
+        "metadata_keys": {"dcp_leadagencyforenvreview": ("name", "accountid")},
+    },
+    "current_milestone": {
+        "metadata_field_names": ["dcp_CurrentMilestone"],
```
[the remark this answers is not in the document]
Comment (SashaWeinstein) [reply src/recode_id.py:34]: this field name is set in the CRM

File: src/recode_id.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+from .client import Client
+
+from .util import create_logger
+from . import CLIENT_ID, SECRET, TENANT_ID, ZAP_DOMAIN
+
+RECODE_ID_FIELDS = [
+    "primary_applicant",
+    "ceqr_leadagency",
+    "current_milestone",
+    "current_envmilestone",
```
[the remark this answers is not in the document]
Comment (SashaWeinstein) [reply src/recode_id.py:15]: yes, most of the values in these fields in the source data are null. you can check this in the postgres database

File: src/recode_id.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        return id_val
+    else:
+        human_readable = field_dict[metadata_hr_key]
+        if (field_dict[metadata_id_key] != id_val) and (
+            field_dict[metadata_hr_key] != id_val
+        ):
+            message = f"Mismatch between {field_dict[metadata_id_key]}/{field_dict[metadata_hr_key]} and {id_val} for field {local_fieldname}"
+            logger.info(message)
+
+        logger.info(f"assinging {human_readable} to field {local_fieldname}")
```
[the remark this answers is not in the document]
Comment (SashaWeinstein) [reply src/recode_id.py]: addressed in new commit

**Comment by mbh329** [comment]
outputs on DO look good as well

**Comment by td928** [comment]
There is another type of error for making a request other than the noted 401 error the which is stopping me from testing the entire dataset. But this could either be an issue with 1. my own home network 2. possibly vscode for making the request(?). I am not sure we should be too concerned about this if the github actions are successful. But just making a note for the future, if we ever run into other network issue than the 401 error. 

Screen Shot 2022-06-08 at 4 55 19 PM

**Comment by SashaWeinstein** [comment]
Oh yea timeout errors are just when it takes too long for the API to respond, seems like a network issue

**Review by td928 — APPROVED** [review]
Overall really extensive and impressive work. I like it not only accomplishes the task but it is capable of demonstrating its success by creating the logs.

## functionality 
I tested with smaller sample locally for the process with good results. Also helps me understand the step-by-step process taking place within the code. I don't claim to have a line by line understanding now but at least can appreciate each of the module. 

## qaqc
The way I checked the final results was to open up the spreadsheet in Excel and do a simple filtering on the four columns recoding was done. I feel very confident about all except the primary applicant column because the number of unique value for the column is simply much higher which make a simple filter and eyeballing less reliable. I would love to hear about other qaqc methods.

**Pull request merged**